### DISCUSSION TITLE
Remove "Purge all caches" button

### DISCUSSION THREAD

**Discussion by dimisa-RUAdList** [uBlockOrigin/uBlock-issues#3022] [General] [answered]
https://github.com/gorhill/uBlock/commit/bd7ce412244691290ca134aa344b831aa8f41c53

Previously, with the clear caches button, the user could pull changes within 15 minutes after committing. Now, in the worst case scenario, pulling changes can take up to 5 hours of waiting until the next diff patch cycle occurs.

**Comment by gorhill** [comment]
Click the clock of the list to update is much more friendly to remote server than forcing an update of all lists, and it's even easier for users now, clicking the clock causes the list to update immediately.

You can also tell them to click an [update link like this](https://ublockorigin.github.io/uAssets/update-lists.html?listkeys=RUS-0).

**Reply by dimisa-RUAdList** [reply]
This does not work.

<details>
<summary>Screen</summary>

![upd1](https://github.com/uBlockOrigin/uBlock-issues/assets/20126984/a2d90018-c3bd-4d4f-b33f-2ece1d10b2dd)
![upd2](https://github.com/uBlockOrigin/uBlock-issues/assets/20126984/14084082-e1e8-4dfe-8d87-80b68dcf7a63)

</details>

**Reply by gorhill** [reply]
Those links are 404.

**Reply by dimisa-RUAdList** [reply]
Ok, I added a rule, and I'll check how long it will take to pull it out: https://github.com/easylist/ruadlist/commit/1f70ea0e9819eb9dc59cdd59c3debbc0730b2f67

And I corrected the pictures.

**Reply by dimisa-RUAdList** [reply]
@gorhill Can you see the pictures now?

**Reply by dimisa-RUAdList** [reply]
Now I've removed the rule: https://github.com/easylist/ruadlist/commit/361355989892b042f6cf54cca454b6c183cd4ab2

I'm guessing this change will only be available for pull after 5 hours.

**Reply by gorhill** [reply]
You can force an update from origin with [this sort of link](https://ublockorigin.github.io/uAssets/update-lists.html?listkeys=RUS-0&manual=231214). It's new in last dev build, so hopefully this works.

**Reply by dimisa-RUAdList** [reply]
The fact of the matter is that clicking on this link does not lead to an update. And click on the clock icon too.

**Reply by gorhill** [reply]
My bad, the link was improperly crafted, you can try again now.

**Reply by gorhill** [reply]
You can also shift-click the clock.

**Reply by gorhill** [reply]
I found an issue with the link approach, so for now use the shift-click clock approach.

**Reply by dimisa-RUAdList** [reply]
Unfortunately, even clicking on the clock icon while holding down the Shift key still does not lead to an update. At least changes to the filter are not pulled out.

**Reply by gorhill** [reply]
Yes, there is an issue with both ways, I am trying to track down the cause.

**Reply by gorhill** [reply]
[One of the sublists](https://raw.githubusercontent.com/easylist/ruadlist/master/AWRL-non-sync.txt) of RUAdList has a `Last modified:` dated `13 Oct 2023 13:30 UTC`, and there are no other valid `Last modified:` in other lists so uBO kept seeing the whole list as older than the ones from CDNs, so it skipped over it.

**Reply by dimisa-RUAdList** [reply]
Fix: https://github.com/easylist/ruadlist/commit/e3cbfb2e94e24e424def014d12556bc6f863fad6

**Reply by dimisa-RUAdList** [reply]
Yes, everything is fine now!

**Reply by gorhill** [reply]
I will still push a fix to ignore `Last modify` field when pulling resource from origin server as opposed to when pulling from a CDN server.

**Reply by dimisa-RUAdList** [reply]
But in the filter header after the update for some reason:
```
! Title: RU AdList for uBO only
! Last modified: %timestamp%
```

**Reply by dimisa-RUAdList** [reply]
Maybe I should remove the line from the sublist header altogether: `! Last modified: %timestamp%`?

**Reply by gorhill** [reply]
> for some reason

What is the issue?

**Reply by dimisa-RUAdList** [reply]
In EasyList it looks like this: `! Last modified: Thu, 14 Dec 2023 16:52:05 +0000`

**Reply by gorhill** [reply]
Do not remove it, this will prevent the GitHub workflow to patch the proper timestamp. It's a placeholder to be filled in later with the scripts which publishes the minified list.

**Reply by dimisa-RUAdList** [reply]
I mean sublist header:
https://github.com/easylist/ruadlist/blob/master/css-fixes-experimental.txt
https://github.com/easylist/ruadlist/blob/master/js-fixes-experimental.txt

**Reply by gorhill** [reply]
> In EasyList

Easylist has a script somewhere which assemble the lists [from templates](https://github.com/easylist/easylist/blob/master/easylist.template). You need to ask them if this can be done for ruadlist repo.

**Reply by gorhill** [reply]
> I mean sublist header

Yes, these can be removed if the lists are never going to be used standalone.

**Reply by dimisa-RUAdList** [reply]
I always have them connected separately, this is convenient for debugging. But I don't think any users do the same.

**Reply by dimisa-RUAdList** [reply]
I updated to version 1.54.1b12, but it still requires me to press the Shift key to update successfully.

**Reply by gorhill** [reply]
It's by design, to ensure people don't forfeit the benefit of differential updates. With shift-click, the list will next update in a few days instead of few hours.

**Comment by partingscientist** [comment]
@gorhill Just making sure, if I want to update resources from `userResourcesLocation`, I can still click purge from the devtools and close my browser session to force auto-update on next start while maintaining diff update as before, no?

**Reply by gorhill** [reply]
> if I want to update resources from `userResourcesLocation`

Clicking on the clock of one list doesn't work?

**Reply by partingscientist** [reply]
There is no way I would have know that the update of `userResourcesLocation` is tied to the update of any list unless I scrutinise the source code (unless I missed it in the wiki).

> With shift-click, the list will next update in a few days instead of few hours.

So clicking the clock (without shift) is equivalent to forcing an auto-update, no (thus maintaining diff update)? Also, is the behaviour of the purge button in the devtools identical to the purge button before 1.54.b11?

**Reply by gorhill** [reply]
The "Purge all caches" button in dev tools is for development purpose only, it's equivalent to shift-click "Purge all caches" the old button.

> There is no way I would have know that the update of userResourcesLocation is tied to the update of any list

The [documentation says](https://github.com/gorhill/uBlock/wiki/Advanced-settings#userresourceslocation):

> Purging the cache of 'uBlock filters' will also purge the cache of the built-in resource file -- and hence force a reload of user-specified resources if any.

But this needs correction, it's really any list currently, not just _uBlock filters_.

**Reply by partingscientist** [reply]
> Purging the cache of 'uBlock filters'

Yeah I have been interpreting this as "Using the purge cache button", thank you for answering.

So is this correct, by the way?

> So clicking the clock (without shift) is equivalent to forcing an auto-update, no (thus maintaining diff update)?

**Reply by gorhill** [reply]
Yes, and this should force a fetch of resources listed in `userResourcesLocation`.

**Reply by partingscientist** [reply]
Thank you.

**Comment by dimisa-RUAdList** [comment]
@dportvine Does it work for you?

**Reply by dportvine** [reply]
Yes it works for me. I see the last change..

<details>
<summary>Screenshot(s)</summary>

![2023-12-15_084559](https://github.com/uBlockOrigin/uBlock-issues/assets/111344219/cffc05ff-9047-40b0-b180-ba5b97a12ea7)
</details>

**Comment by dimisa-RUAdList** [comment]
To summarize, we can conclude that now the control panel interface does not have the ability to update filters to the current state. A simple click on the clock icon causes the filter to be updated to the latest diff patch, which may be a 5-hour outdated version. Moreover, such a click must be made for each filter separately.

The ability to press the shift button is not mentioned anywhere, and even if you use it, you still need to do this for each filter separately.

Yes, when reporting a problem using the chat icon, which has a button to update all filters, they can be updated at the same time. But this option can hardly be considered user-friendly, since it is not obvious and it is impossible to get there directly from the control panel.

I think this is too drastic a decision.

**Reply by dportvine** [reply]
Waiting 5 hours to get a rule added 50 minutes ago is a long time. Is it possible to reduce the interval?
Okay, this works for me. https://ublockorigin.github.io/uAssets/update-lists.html?listkeys=RUS-0&manual=231214

**Reply by Yuki2718** [reply]
I think it makes sense to reinstate these buttons under filterAuthorMode.

**Reply by dimisa-RUAdList** [reply]
How about putting a button to unconditionally update filters on the support tab?

**Reply by gorhill** [reply, marked as answer]
Filter lists should never need to be updated manually for the vast majority. The issue here is that "Purge all cache"-"Update now" is a meme but nobody is willing to foot the server bills should we be kicked out of one or more of those CDNs uBO uses.

This is the reality:

- No, you do not need to update lists all the time without actual valid reasons
- Differential updates ensure people will be very close to up to date lists at all time without burdening servers
- In the event a specific fix has been added and a user want to get it immediately, we give them a crafted link to make it easy for them
- There is no need to update _all_ lists to absolute latest versions, this is wasted bandwidth and kills differential update cycles, which will takes days to come back, i.e. the user is worse off doing so.

> The ability to press the shift button is not mentioned anywhere, and even if you use it, you still need to do this for each filter separately.

In what reality-based scenario does one need to do this for every lists? People report filter issues, they are fixed, we give them a link to update only the one list which has changed. If I put pack another way to do the same as "Purge all cache"-"Update now", it will be abused the same way because people out there are not concerned with [CDN costs](https://github.com/uBlockOrigin/uBlock-issues/discussions/2958).

The changes here ensure that the action of updating lists is driven by actual filter issues, not by how user feels when their lists is not perfectly up to date to the minute even when there are no filter issues.

**Reply by MasterKia** [reply]
@gorhill Side-note: What is the minimum supported `Expires` time for a filter list?
Last I recall, it was 12 hours. Is it 6 hours now?

Are these correct?
```adb
! Expires: 1 days
! Expires: 12h
! Expires: 6h
```

**Reply by Yuki2718** [reply]
Is `12h` correct? I thought it should be `12 hours`.

**Reply by gorhill** [reply]
Minimum is 4h, and `12h` should work, no need for full `hours`.

**Reply by epignatelli** [reply]
> - No, you do not need to update lists all the time without actual valid reasons
> - Differential updates ensure people will be very close to up to date lists at all time without burdening servers

This is not true. 
For youtube, people found that purging and updating all caches was the only way for uBlock to work.

The update is, most likely, to cause users to migrate off uBlock for a competitor.

**Reply by gorhill** [reply]
What specific issue are you having at the moment that you need a "Purging all caches" button?

"Purging all caches" was not a magic wand solving all filtering issues, there was no point using it if filter solutions had not been identified and added to filter lists. If ever a new filter solution is added and needs to be pushed to users, we provide click-to-update links in [the official Youtube thread](https://github.com/uBlockOrigin/uAssets/issues/20586).

Currently filter lists update every 5-6 hours, and when you open the "Filter lists" pane they should always be up to date with most recent filters. We provide click-to-update links as needed when solving filter issues -- they are guaranteed to work and are more user-friendly, unlike blind "Purge all caches"/"Update now" which is just wishful thinking and putting uBO at risk of being kicked out of the CDNs hosting its lists.

Wanting stuff without being concerned is easy. Not so much [when you have to pay for it](https://github.com/uBlockOrigin/uBlock-issues/discussions/2958).

**Reply by rorybd** [reply]
I agree that bandwidth was an issue and fully support differential updates!

However, it appears that something may be wrong with the click-to-update links for some users. I was recently unable to bypass pop-ups without fully resetting the extension, which I'd like not to do so as not to incur additional costs.

**Reply by gorhill** [reply]
> something may be wrong with the click-to-update links for some users

Yes, we are looking into it, but typically people who have issues should not presume the issue is that they don't have up to date lists, with all the latest improvements, uBO is staying ahead of YT changes (though this may change at any time, but currently we are ahead). Youtube anti-ad block can be triggered for plenty other reasons, hence why we always need troubleshooting info -- asking back the "Purge all caches" button is not the solution, we first need to investigate that the issue is not one of countless external factors. When the issue is one of these other factors, "Purge all caches" just consume bandwidth and issue is still present.

**Reply by rorybd** [reply]
I fully agree! I merely wanted to bring this up in case it was not already known. My apologies if this was not the appropriate place for that.

**Reply by gorhill** [reply]
We have fixed issues with the status page, it should work now, you are guaranteed to fetch the latest fixes: <https://drhyperion451.github.io/does-uBO-bypass-yt/>.

This is the official way to update for Youtube -- as long as all the external factors which can also cause detection are being taken care of by following the the other steps at <https://github.com/uBlockOrigin/uAssets/issues/20586>.

**Reply by nonstopfox** [reply]
whether the purge all cache button was intended to or not, it did fix issues some Ublock users (myself included) experienced with youtube. i definitely wouldn't want the Ublock project to run into any walls due to server costs and overuse of a feature in a way it wasn't intended to be used but you can't just say it wasn't a magic wand fix when it did in fact fix the problem, intentionally or not. 

since the removal of the purge button i've been experiencing the same issues as usual with no better solution than what was there before, until i found this link https://drhyperion451.github.io/does-uBO-bypass-yt/ 

updating uBO quick fixes does solve the problem for me the same way the purge all caches option did and this seems to be the preferred solution by the ublock team according to the one who provided the link to me. so to anyone having this same issue who came here for an answer, try the solution provided by the ublock team in the link. 
i'm happy to see a better solution to the problem and hope you don't run into too much trouble explaining why the change was needed to the users. best wishes to the ublock team and thanks for your time and work as usual.

**Reply by Yuki2718** [reply]
You never know if the button "fixed" issues or not unless you technically investigated the cause. The recent slowdown fuss with Adblock demonstrated how people tend to mis-assign cause of problems. In the field of science it is well known as confusion of causal relationship with correlation. The issue disappeared after one did the steps only means correlation which can very well be coincidence - in particular sometimes the warning disappers just with refreshing. We learned many people didn't correctly eliminated external factors, or unable to do so. People just learned, like the Pavlovian dog, the warning disappears after they did the step even when there is a real cause. We say not a magic wand simply because it's not something to fix the issue by itself, but is to only fetch the fix added in upstream filters. It is no more needed with introduction of differential update, uritransform, and the update link. We made a mistake yesterday which caused the warning to some user, as have done in past too, but the affected user should not be too many considering EasyList is also diff-updated.

**Comment by stephenhawk8054** [comment]
How should we modify the issue template? It's now still having "Purge all caches" button

**Reply by gorhill** [reply]
We could remove the generic template ant direct people to use the chat icon.

**Reply by stephenhawk8054** [reply]
I mean the line

- [ ] I forced an update of my filter lists. (Click the ["Purge all caches"](https://github.com/gorhill/uBlock/wiki/Dashboard:-Filter-lists#purge-all-caches) button while holding the 'Shift' key, then click the ["Update now"](https://github.com/gorhill/uBlock/wiki/Dashboard:-Filter-lists#update-now) button.)

when creating the issues via chat icon. I think we don't need it any more right? Given there's already an `Update` button before users creating the issue.

**Reply by dimisa-RUAdList** [reply]
What if the user does not use the chat icon on the panel and immediately writes here? ~ > https://github.com/uBlockOrigin/uAssets/issues/new/choose

**Reply by krystian3w** [reply]
Maybe need implement these:
```adb
github.com##:matches-path(/\/uBlockOrigin\/uAssets[\w\W]+\/new/) .timeline-comment > div:nth-child(-n+3)
```
